### PR TITLE
fix(gc): update repodb when gc'ing manifests

### PR DESCRIPTION
**Pull request by peusebiu**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/storage/gc/gc.go [hunk 1]
```diff
@@ -0,0 +1,687 @@
...
+
+		return err
+	}
+
+	gc.log.Info().Msg(fmt.Sprintf("GC successfully completed for %s", path.Join(gc.imgStore.RootDir(), repo)))
+
+	return nil
+}
+
+func (gc GarbageCollect) repo(repo string) error {
```
Comment (andaaron) [inline pkg/storage/gc/gc.go:82]: Needs a better name

File: pkg/storage/gc/gc.go [hunk 2]
```diff
@@ -0,0 +1,687 @@
...
+func (gc GarbageCollect) RunPeriodically(interval time.Duration, sch *scheduler.Scheduler) {
+	generator := &GCTaskGenerator{
+		imgStore: gc.imgStore,
+		gc:       gc,
+	}
+
+	sch.SubmitGenerator(generator, interval, scheduler.MediumPriority)
+}
+
+func (gc GarbageCollect) Repo(repo string) error {
```
Comment (andaaron) [inline pkg/storage/gc/gc.go:66]: Needs a better name

File: pkg/storage/gc/gc.go [hunk 3]
```diff
@@ -0,0 +1,687 @@
...
+	gc.log.Info().Msg(fmt.Sprintf("executing GC of orphaned blobs for %s", path.Join(gc.imgStore.RootDir(), repo)))
```
Comment (andaaron) [inline pkg/storage/gc/gc.go:67]: I think there are too many non-debug logs. Maybe make them debug logs.

File: pkg/meta/boltdb/boltdb.go [hunk 4]
```diff
@@ -404,6 +404,47 @@ func (bdw BoltDB) GetReferrersInfo(repo string, referredDigest godigest.Digest,
 	return referrersInfoResult, err
 }
 
+func (bdw *BoltDB) RemoveRepoReference(repo, reference string, manifestDigest godigest.Digest) error {
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/meta/boltdb/boltdb.go:407]: gc logic should skip manifest removal if there are conflicts, I have a check for that.

File: pkg/storage/gc/gc.go [hunk 5]
```diff
@@ -0,0 +1,687 @@
...
+			if image.Subject != nil {
+				referenced[desc.Digest] = true
+			}
+		}
+	}
+
+	return nil
+}
+
+func (gc GarbageCollect) blobs(repo string, index ispec.Index,
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/storage/gc/gc.go:410]: I was thinking that gc.blobs() should be self explanatory :D, I'll change, thanks.

File: pkg/storage/imagestore/imagestore.go [hunk 6]
```diff
@@ -1497,6 +1501,48 @@ func (is *ImageStore) DeleteBlob(repo string, digest godigest.Digest) error {
 	return is.deleteBlob(repo, digest)
 }
 
+/*
+CleanupRepo removes blobs from the repository and removes repo if flag is true and all blobs were removed
+the caller function SHOULD lock from outside.
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/storage/imagestore/imagestore.go:1506]: yes

File: pkg/meta/boltdb/boltdb.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (andaaron) [reply pkg/meta/boltdb/boltdb.go:407]: 1. We can't assume whoever will se this method will know not to use it outside the scope of GC, and will make sure  the GC will do those checks when he refactors.
2. The method states it removes a refence, not that it's removing everything else related to the digest too.
To me it looks like a data loss scenario waiting to happen if we keep it like this.

File: pkg/storage/gc/gc.go [hunk 7]
```diff
@@ -0,0 +1,687 @@
...
+			Msg("gc: failed to stat blob")
+
+		return false, err
+	}
+
+	if modtime.Add(delay).After(time.Now()) {
+		return false, nil
+	}
+
+	log.Info().Str("repository", repo).Str("digest", digest.String()).Msg("perform GC on blob")
```
[the remark this answers is not in the document]
Comment (peusebiu) [reply pkg/storage/gc/gc.go]: need to fix it.

File: pkg/meta/hooks.go [hunk 8]
```diff
@@ -109,7 +109,7 @@ func OnDeleteManifest(repo, reference, mediaType string, digest godigest.Digest,
 			manageRepoMetaSuccessfully = false
 		}
 	} else {
-		err = metaDB.DeleteRepoTag(repo, reference)
+		err = metaDB.RemoveRepoReference(repo, reference, digest)
```
Comment (peusebiu) [inline pkg/meta/hooks.go:112]: @laurentiuNiculae can you review this please?

Comment (peusebiu) [reply pkg/meta/hooks.go:112]: I'm trying to fix the bug where we don't clean digests inside RepoMetadata

File: pkg/storage/gc/gc.go (hunk 3 above)
Comment (peusebiu) [reply pkg/storage/gc/gc.go:78]: I preserved these ones, and I put the rest on debug. These ones are checked in our tests.

**Review by andaaron — DISMISSED** [review]

File: pkg/meta/boltdb/boltdb.go [hunk 9]
```diff
@@ -404,6 +404,79 @@ func (bdw BoltDB) GetReferrersInfo(repo string, referredDigest godigest.Digest,
...
+		}
+
+		if !common.ReferenceIsDigest(reference) {
+			delete(repoMeta.Tags, reference)
+		} else {
+			// find all tags pointing to this digest
+			tags := []string{}
+			for tag, desc := range repoMeta.Tags {
+				if desc.Digest == reference {
+					tags = append(tags, tag)
```
Comment (laurentiuNiculae) [inline pkg/meta/boltdb/boltdb.go:444]: I think we could directly delete the tag here but it's ok this way as well
```suggestion
					delete(tags, tag)
```

**Review by laurentiuNiculae — DISMISSED** [review]

**Review by andaaron — APPROVED** [review]

**Review by rchincha — APPROVED** [review]
lgtm

**Pull request merged**